python/src/iceberg/types.py

```diff
@@ -15,157 +15,1219 @@
+                Tuple[str, Type[Union[IcebergType, "Instance"]], Any],
+            ]
+        ],
+        base_type: type = IcebergType,
+        doc: str = "",
+        meta_doc: str = "",
     ):
-        self._is_optional = is_optional
-        self._id = field_id
-        self._name = name
-        self._type = field_type
-        self._doc = doc
+        """
+        facilitates generating generic type factories such as `List` e.g. `List[Integer]`
+        in the spirit of builtin `type` e.g.:
+            def Robot_init(self, name):
+                self.name = name
 
-    @property
-    def is_optional(self) -> bool:
-        return self._is_optional
+            Robot2 = type("Robot2",
+                          (),
+                          {"counter":0,
+                           "__init__": Robot_init,})
 
-    @property
-    def is_required(self) -> bool:
-        return not self._is_optional
+        Args:
+            name: the name of the generic type used in repr and str
+            attributes: list of generic attribute names and types - must provide at least a name type or both for each attribute
+            base_type: class for generic to inherit from
+            doc: docstring of specfied generic e.g. List[Integer]
+            meta_doc: docstring of unspecified generic e.g. List
 
-    @property
-    def field_id(self) -> int:
-        return self._id
+        Examples:
+            >>>List = generic_class("List", [("type", IcebergType)], PrimitiveType)
+            >>>List[Integer]
+            List[type=Integer]
+        """
+        attributes = [a if len(a) == 3 else (*a, None) for a in attributes]
+        attribute_names = [a[0] for a in attributes]
+        attribute_types = [a[1] for a in attributes]
+        attribute_defaults = dict(zip(attribute_names, [a[2] for a in attributes]))
 
-    @property
-    def name(self) -> str:
-        return self._name
+        def get_generic(cls, attrs):
+            attrs = attrs if isinstance(attrs, tuple) else (attrs,)
 
-    @property
-    def type(self) -> Type:
-        return self._type
+            if len(attrs) > len(attributes):
+                raise TypeError(
+                    f'Too many generic parameters provided. Expected {len(attribute_names)} parameter(s) of subtypes of ({",".join(repr(i) for i in attribute_types)}). Provided {attrs}'
+                )
 
-    def __repr__(self):
-        return (
-            f"NestedField(is_optional={self._is_optional}, field_id={self._id}, "
-            f"name={repr(self._name)}, field_type={repr(self._type)}, doc={repr(self._doc)})"
+            kwargs = dict(zip(attribute_names, attrs))
+            for k, v in attribute_defaults.items():
+                if kwargs.get(k) is None:
+                    if v is None:
+                        raise TypeError(
+                            f"Generic attribute `{str(k)}` on {name} requires a value and has no default."
```

Comment:
Yeah, it's meant to ensure that you fully specify what the type of a generic is so for example if you try `Decimal[32]` you'll get `TypeError: Generic attribute `scale` on Decimal requires a value and has no default.` since you did not specify the scale